%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Simple Sectioned Essay Template
% LaTeX Template
%
% This template has been downloaded from:
% http://www.latextemplates.com
%
% Note:
% The \lipsum[#] commands throughout this template generate dummy text
% to fill the template out. These commands should all be removed when 
% writing essay content.
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%----------------------------------------------------------------------------------------
%	PACKAGES AND OTHER DOCUMENT CONFIGURATIONS
%----------------------------------------------------------------------------------------

\documentclass[12pt]{article} % Default font size is 12pt, it can be changed here
% Margin specification
% Check http://en.wikibooks.org/wiki/LaTeX/Page_Layout for more info
\usepackage[margin = 1in]{geometry}
\usepackage[nottoc,notlof,notlot,numbib]{tocbibind}

% Some misc and math packages
% Check http://en.wikibooks.org/wiki/LaTeX/Mathematics for more info
\usepackage{fancyhdr}
\usepackage{manfnt}
\usepackage{pgf}
\usepackage{amsmath,amsthm,amssymb,natbib,graphicx}
\usepackage{amsfonts}
\DeclareMathAlphabet{\mathpzc}{OT1}{pzc}{m}{it}
\usepackage{bbm}
\usepackage{float}
\usepackage{mathrsfs} %mathscr{A}
\usepackage{hyperref}
\usepackage{pdfpages}

% Color
\usepackage{color}

% For specifying the counter style of enumerate
\usepackage{enumerate}

\usepackage{geometry} % Required to change the page size to A4
%\geometry{a4paper} % Set the page size to be A4 as opposed to the default US Letter

\usepackage{graphicx} % Required for including pictures

\usepackage{float} % Allows putting an [H] in \begin{figure} to specify the exact location of the figure
\usepackage{wrapfig} % Allows in-line images such as the example fish picture

\usepackage{parskip} % Space between paragraphs

\usepackage{lipsum} % Used for inserting dummy 'Lorem ipsum' text into the template

\linespread{1.5} % Line spacing

\setlength\parindent{0pt} % Uncomment to remove all indentation from paragraphs

\graphicspath{{./Pictures/}} % Specifies the directory where pictures are stored

\newcommand{\iid}{\stackrel{\mathrm{iid}}{\sim}}

% Page style definition
\rfoot{\today}
\begin{document}


%------------------------------------------------
\begin{titlepage}

\newcommand{\HRule}{\rule{\linewidth}{0.5mm}} % Defines a new command for the horizontal lines, change thickness here

\center % Center everything on the page
 
%----------------------------------------------------------------------------------------
%	HEADING SECTIONS
%----------------------------------------------------------------------------------------

\textsc{\LARGE Johns Hopkins University}\\[1.5cm] % Name of your university/college
\textsc{\Large Computational Genomics}\\[0.5cm] % Major heading such as course name
\textsc{\large EN 600.639}\\[0.5cm] % Minor heading such as course title

%----------------------------------------------------------------------------------------
%	TITLE SECTION
%----------------------------------------------------------------------------------------

\HRule \\[0.4cm]
{ \huge \bfseries Augmenting Phylogenetic Classification of Metagenomic Reads}\\[0.4cm] % Title of your document
\HRule \\[1.5cm]
 
%----------------------------------------------------------------------------------------
%	AUTHOR SECTION
%----------------------------------------------------------------------------------------

\begin{minipage}{0.4\textwidth}
\begin{flushleft} \large
\emph{Authors:}\\
Shuya \textsc{Chu}, \\  Stephen \textsc{Cristiano}, \\ Bing \textsc{He},\\  Zhou \textsc{Ye} % Your name
\end{flushleft}
\end{minipage}
~
\begin{minipage}{0.4\textwidth}
\begin{flushright} \large
\emph{Instructor:} \\
Dr. Ben \textsc{Langmead} % Supervisor's Name
\end{flushright}
\end{minipage}\\[4cm]

% If you don't want a supervisor, uncomment the two lines below and remove the section above
%\Large \emph{Author:}\\
%John \textsc{Smith}\\[3cm] % Your name

%----------------------------------------------------------------------------------------
%	DATE SECTION
%----------------------------------------------------------------------------------------

{\large \today}\\[3cm] % Date, change the \today to a set date if you want to be precise

%----------------------------------------------------------------------------------------
%	LOGO SECTION
%----------------------------------------------------------------------------------------

%\includegraphics{Logo}\\[1cm] % Include a department/university logo - this will require the graphicx package
 
%----------------------------------------------------------------------------------------

\vfill % Fill the rest of the page with whitespace
 \end{titlepage}
%------------------------------------------------
\vspace{5 ex}
\section{Abstract}
In recent years, metagenomics has gained increased interest in the scientific community. With relevance to  public health, agriculture, biofuels, and marine biology [REFERENCE], it is important to understand the taxonomic composition of metagenomic samples and its functional impact it their respective ecosystems. In this project, we use simulated data to explore a variety of methods to taxonomically classify metagenomic reads to their correct position in a phylogenetic reference set. RESULTS


\section{Introduction}
The recent expansion of next generation sequencing technologies have greatly benefited our ability to study microbial communities [Reference]. Metagenomics, defined in 2005 as ``The application of modern genomics techniques to the study of communities of microbial organisms directly in their natural environments, bypassing the need for isolation and lab cultivation of individual species'' [\cite{Chen:2005dg}], has seen a surge of research efforts in recent years in diverse fields including public health, agriculture, biofuels, and marine biology, and efforts such as the Human Microbiome Project generating large quantities of metagenomic data. Despite this, computational methods are still needed for analyzing and making sense of these data in an accurate but efficient matter. 
\par
One important challenge is, given a set of metagenomic reads, to accurately classify the taxonomic composition of the reads. For example, a metagenomic human gut sample will contain many different microbial species. We wish to be able to classify each read to the correct phylogenetic location in a reference set and determine the presence and abundance of different microbial species in the sample. Adding to the challenge, many of the species will be closely related (belonging to the same genus, family etc) and in a microbial sample, many of the reads make come from \emph{de novo} genomes, \emph{ie} from species that have never been sequenced before and unknown to the scientific community. 
\par
In this project, we propose, implement, and compare three methods for the classification of metagenomic reads. In particular, we provide a Bloom filter approach, an approach based on Interpolated Markov Models, and an approach based using Bowtie to perform local alignment to reference genomes and classify using a novel algorithm. While similar methods exist, we also consider a number of ways in which we can improve upon these.
\par
To perform our analyses, we obtain reference genomes from RefSeq, a large collection of genomes maintained by the National Center for Biotechnology Institute (NCBI) [\cite{Pruitt:2009dg}] 
\section{Related work}
A large body of work has already been done on the phylogenetic classification of metagenomic reads. It's important to note that many of the early methods are only accurate for large sequence reads and fail at correctly classifying the short ($ \sim100$ bp) reads generated by next generation sequencers [\cite{Brady:2009dg}]. We briefly review a few of the more recent metagenomic classification methods.
\subsection{PhymmBL}
PhymmBL [\cite{Brady:2009dg}] uses an Interpolated Markov Model to phylogenetically classify genomic reads to genomes in RefSeq, a large collection of genomes. It is capable of performing classification at different levels of the phylogentic tree (species, genus, family etc). In addition to classification using an IMM, PhymmBL also uses BLAST for local alignment to classify. By applying a linear function of the scores from the IMM and BLAST, the authors report greater classification accuracy than either the IMM or BLAST alone. Though very accurate, PhymmBL suffers from extremely slow running time, due to using BLAST for local alignment. 

\subsection{FACS}
Fast Accurate Classification of Sequences (FACS) [CITATION] transforms the reference genomes to Bloom filters and query Bloom filters for exact matches. Bloom filters are an compact hash-based data structure which are used to quickly determine whether an element is a member of a set or not. In a trade-off for compactness and speed of look-up, Bloom filters come with a risk of giving false positives which must be controlled. FACS is implemented in PERL and achieves high speed, but is limited to using Bloom filters $<312 MB$ and does not have a scoring system optimized for classification. In a comparison of classification software, other authors have reported an inability to reproduce the results reported in the FACS paper [\cite{Bazinet:2012dg}].

\subsection{MetaPhlAn}
MetaPhlAn (Metagenomic Phylogenetic Analyis) [HUTTENHOWER] is a popular tool for metagenomic classification. 

\subsection{Bowtie}
Bowtie [\cite{Langmead:2009dg}] utilizes a Burrow-Wheeler Transform and an FM index to perform extremely fast alignment of short DNA reads to a reference genome. By performing local alignment of the reads to each of the reference genomes from a phylogenetic tree, and using an appropriate scoring system, Bowtie can be adapted for classification of metagenomic reads. 

\section{Methods and software}
We obtain our reference genomes from RefSeq, obtained from the NCBI. Due to the vast size of RefSeq, we initially considered only the phylum \emph{Proteobacterium}, but even that exceeded our memory limits. For this reason, we decided to use comparisons of the methods for our reference set the genomes from \emph{Escherichia coli} (strain K-12 substain. DH10B) and  {\bf[BDELLO GENOME]}. To simulate our data, we use the software MetaSim [\cite{Richter:2008dg}] which will generate metagenomic reads into a FASTA file given a set of reference genomes.
\par
As the main purpose of the Bowtie method was to develop an algorithm that will not have align to every genome in a reference set, we also considered 15 genomes from \emph{Proteobacterium} but the same simulated reads. However, the comparisons with the other methods were still conducted only on the two previously mentioned reference genomes.
\subsection{Bloom filter}
Our first method was to create a Bloom filter for each reference genome and query the reads to them. 

\subsection{Improvement with Bowtie2}
Bowtie2 is an ultrafast aligner using FM index. It is particularly good at aligning reads of about 50 up to 100s or 1,000s of characters. Since our simulate testing data is around 100-200 long, using Bowtie2 will lead to fast aligning.
\par
The idea for improvement is by skipping some unnecessary alignment between the test data and reference, identifying the species for test data will be much more fast. Like what we did in the Boyer-Moore method in class, we use results from comparisons to skip future alignments that definitely won't match. To be more specific,  if we got a really low score when aligning the test data with one species, it means the sample is definitely not from this species, thus we can skip the whole species from the same genus. For example, if we know this sample definitely does not belong to \emph{Canis}, instead of checking whether it's \emph{Canis lupus}, we can check whether it's \emph{Cuon alpinus} from the genus of \emph{Cuon}.
\par
First we construct a tree structure to describe the taxonomy relation between the reference in python. For each node, node.data represent directory path of this reference, parent and children is directory path for its {\bf XXXX}. 
\par
Then by running Bowtie2, we will have scores which describe the similarity between the sample data and reference. If scores is really low, in other words, we have enough confidence to say the sample is not from this species, we skip the whole genus this reference species belongs to. If scores is not that bad, we will go through all the species in this genus to determine which species it has most similarity with.  
\par
Threshold here is very important. We didn't come up with a fixed expression to calculate it by the due date. We extract 30 species from 20 genus from the reference set and set the threshold after observing the result from those alignments. Also, we used the average score from the scores of 1000 test sequence in the .fna file.
\par
In order to run the code, please change the value of 'projectpath' to the actual path that 'AAA.zip' you extract to. And also, make sure inside the 'projectpath', there is a /data/index folder and a /bowtie2-2.1.0 folder.

\section{Results}
After running \verb|bowtie_with_skipping.py| and \verb|without_skipping|, we can tell with {\bf X species in X genus, skipping method is X times faster}.

\section{Conclusion}
In our project, we implemented and compared three methods for the phylogenetic classification of metagenomic reads. Under our simulated data, 
\par
For future work, we would ideally run our methods on reference genomes from the full RefSeq and using simulated reads from a larger clade in the tree. While we aimed to do so for our project, we found that we found that the limitations on cluster we were using were too stringent and we had to vastly scale down the scale of our reference set. In particular, we believe that the IMM method will improve as we scale up the data and the number of reference genomes. We also would like to scale our comparisons to real data instead of simulated reads, especially a dataset that has been well studied and we have some idea of what the true classifications should be. It's also important to compare our methods to the existing methods. While we believe we sufficiently improved upon existing methods, we were unable to get existing software running on the cluster, aside from FACS. 
\par
Other things we would like to test is the ability of our methods to detect \emph{de novo} reads. A real metagenomics sample will have reads from many bacteria which have never been discovered, which we would like our methods to classify to the correct genus, but identify as being from an unknown genome. One easy way we can explore this is by simulating reads from a reference set of genomes, and then classify against the reference set multiple times using a cross validation procedure with one reference genome left out from each classification.
\par
In summary, we identify three methods for phylogenetic classification for metagenomics. While our initial results  look very promising, we need to scale up the size of our simulations and set of reference genomes to fully quantify the accuracy, sensitivity, and specificity of these methods. We remain hopeful that our methods will perform well on a more realistic simulation and can be applied to real metagenomics data. 
%------------------------------------------------

%----------------------------------------------------------------------------------------
%	BIBLIOGRAPHY
%----------------------------------------------------------------------------------------

\begin{thebibliography}{99} % Bibliography - this is intentionally simple in this template
\bibitem[Chen, Pacther (2005)]{Chen:2005dg}
Chen K, Pachter L (2005). 
\newblock Bioinformatics for Whole-Genome Shotgun Sequencing of Microbial Communities. 
\newblock {\em PLoS Comput Biol}, 1(2): e24. doi:10.1371/journal.pcbi.0010024

\bibitem[Brady, Salzberg (2009)]{Brady:2009dg}
Brady A, Salzberg S (2009).
\newblock Phymm and PhymmBL: metagenomic phylogenetic classification with interpolated Markov models.
\newblock {\em Nature Methods} 6, 673 - 676 (2009) 

\bibitem[Pruitt et al (2009)]{Pruitt:2009dg}
Pruitt KD, Tatusova T, Klimke W, Maglott DR. 
\newblock NCBI Reference Sequences: current status, policy and new initiatives. 
\newblock {\em Nucleic Acids Res.} 2009 Jan; 37 (Database issue):D32-36.

\bibitem[Bazinet, Cummings (2012)]{Bazinet:2012dg}
Bazinet AL, Cummings MP (2012)
\newblock A comparative evaluation of sequence classification programs
\newblock {\em BMC Bioinformatics} 2012, 13:92

\bibitem[Stranneheim et al (2010)]{Stranneheim:2010dg}
Stranneheim H, Käller M,  Allander T,  Andersson B,  Arvestad L,  Lundeberg J 
\newblock Classification of DNA sequences using Bloom filters
\newblock {\em Bioinformatics} (2010) 26 (13): 1595-1600.

\bibitem[Langmead et al (2010)]{Langmead:2009dg}
Langmead B,  Trapnell C,  Pop M,  Salzberg S
\newblock Ultrafast and memory-efficient alignment of short DNA sequences to the human genome.
\newblock {\em Genome Biology} (2010) 10:R25

\bibitem[Ricther et al (2008)]{Richter:2008dg}
Richter DC, Ott F, Auch AF, Schmid R, Huson DH 
\newblock MetaSim—A Sequencing Simulator for Genomics and Metagenomics. 
\newblock PLoS ONE (2008) 3(10): e3373. doi:10.1371/journal.pone.0003373
\end{thebibliography}
%----------------------------------------------------------------------------------------

\end{document}